repo/blob/sftp/sftp_storage.go

```diff
@@ -0,0 +1,338 @@
+package sftp
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	psftp "github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+
+	"github.com/kopia/kopia/repo/blob"
+)
+
+const (
+	sftpStorageType      = "sftp"
+	fsStorageChunkSuffix = ".f"
+
+	packetSize = 1 << 15
+)
+
+var (
+	sftpDefaultShards     = []int{3, 3}
+	sftpDefaultKnownHosts = filepath.Join(os.Getenv("HOME"), ".ssh", "known_hosts")
+)
+
+// sftpStorage implements blob.Storage on top of sftp.
+type sftpStorage struct {
+	Options
+
+	conn *ssh.Client
+	cli  *psftp.Client
+}
+
+func (s *sftpStorage) GetBlob(ctx context.Context, blobID blob.ID, offset, length int64) ([]byte, error) {
+	_, path := s.getDirPathAndFilePath(blobID)
+
+	r, err := s.cli.Open(path)
+	if os.IsNotExist(err) {
+		return nil, blob.ErrBlobNotFound
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+
+	// pkg/sftp doesn't have a `ioutil.Readall`, so we WriteTo to a buffer
+	// and either return it all or return the offset/length bytes
+	buf := new(bytes.Buffer)
+	n, err := r.WriteTo(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	if length < 0 {
+		return buf.Bytes(), nil
+	}
+
+	if offset > n || offset < 0 {
+		return nil, errors.New("invalid offset")
+	}
+
+	data := buf.Bytes()[offset:]
+	if int(length) > len(data) {
+		return nil, errors.New("invalid length")
+	}
+
+	return data[0:length], nil
+}
+
+func getBlobIDFromFilename(name string) (blob.ID, bool) {
+	if strings.HasSuffix(name, fsStorageChunkSuffix) {
+		return blob.ID(name[0 : len(name)-len(fsStorageChunkSuffix)]), true
+	}
+
+	return "", false
+}
+
+func makeFileName(blobID blob.ID) string {
+	return string(blobID) + fsStorageChunkSuffix
+}
+
+func (s *sftpStorage) ListBlobs(ctx context.Context, prefix blob.ID, callback func(blob.Metadata) error) error {
+	var walkDir func(string, string) error
+
+	walkDir = func(directory string, currentPrefix string) error {
+		entries, err := s.cli.ReadDir(directory)
+		if err != nil {
+			return err
+		}
+
+		for _, e := range entries {
```

Comment:
This one, I'm not sure how to fix.

I could surely create a common function that adds a `readDir` function argument that is provider specific.

But it would only serve filesystem and sftp, webdav's logic is slightly different.

Reply:
See #56